sb3_contrib/ars/ars.py

```diff
@@ -0,0 +1,276 @@
+        seed: Optional[int] = None,
+        verbose: int = 0,
+        device: Union[th.device, str] = "cpu",
+        _init_setup_model: bool = True,
+        alive_bonus_offset: float = 0,
+        zero_policy: bool = True,
+    ):
+
+        super().__init__(
+            policy,
+            env,
+            learning_rate=0.0,
+            tensorboard_log=tensorboard_log,
+            policy_base=policy_base,
+            policy_kwargs=policy_kwargs,
+            verbose=verbose,
+            device=device,
+            supported_action_spaces=(gym.spaces.Box,),
+            support_multi_env=True,
+        )
+
+        self.n_delta = n_delta
+        self.alpha = get_schedule_fn(alpha)
+        self.sigma = get_schedule_fn(sigma)
+
+        if n_top is None:
+            n_top = n_delta
+        self.n_top = n_top
+
+        self.n_workers = None  # We check this at training time ... I guess??
+
+        if policy_kwargs is None:
+            policy_kwargs = {}
+        self.policy_kwargs = policy_kwargs
+
+        self.seed = seed
+        self.alive_bonus_offset = alive_bonus_offset
+
+        self.zero_policy = zero_policy
+        self.theta = None  # Need to call init model to initialize theta
+
+        if (
+            _init_setup_model
+        ):  # TODO ... what do I do if this is false? am i guaranteed that someone will call this before training?
+            self._setup_model()
+
+    @classmethod  # Override just to change the default device argument to "cpu"
+    def load(
+        cls,
+        path: Union[str, pathlib.Path, io.BufferedIOBase],
+        env: Optional[GymEnv] = None,
+        device: Union[th.device, str] = "cpu",
+        custom_objects: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> "BaseAlgorithm":
+        return super().load(path, env, device, custom_objects, **kwargs)
+
+    def _setup_model(self) -> None:
+        self.set_random_seed(self.seed)
+        self.rng = np.random.default_rng(self.seed)
+
+        self.policy = self.policy_class(self.observation_space, self.action_space, **self.policy_kwargs)
+        self.dtype = self.policy.parameters().__next__().dtype  # This seems sort of like a hack
+        self.theta = th.nn.utils.parameters_to_vector(self.policy.parameters()).detach().numpy()
+
+        if self.zero_policy:
+            self.theta = np.zeros_like(self.theta)
+            theta_tensor = th.tensor(self.theta, requires_grad=False, dtype=self.dtype)
+            th.nn.utils.vector_to_parameters(theta_tensor, self.policy.parameters())
+
+        self.n_params = len(self.theta)
+        self.policy = self.policy.to(self.device)
+
+    def _collect_rollouts(self, policy_deltas, callback):
+        with th.no_grad():
+            batch_steps = 0
+            theta_idx = 0
+
+            # Generate 2*n_delta candidate policies by adding noise to the current theta
+            candidate_thetas = np.concatenate([self.theta + policy_deltas, self.theta - policy_deltas])
```

Comment:
Hmm would it not be more sensible to do all operations with torch tensors instead of creating numpy arrays and moving them back-and-worth torch and numpy?

Reply:
It might yes, I generally prefer to keep things in numpy where possible (because I don't need the gradient information, and multiprocessing with torch tensors has gotchas) but there is no good reason to in this context.